File: Graphs/Silver/P1_2015-Switching_on_the_Lights/P1_2015-Switching_on_the_Lights.py
# Graphs - Breadth First Search (DFS) | Flood fill approach
# T & M: O(n^2 + m)
# Where n is the number of rows and columns in the switches graph and m is the number of switches

from collections import deque

with open("lightson.in", "r") as fin:
	# First line: n = graph side length, m = # of switches
	n, m = map(int, fin.readline().split())
	# For the next m lines map x, y room to room a, b that can be lit
	graph = {}	# 1-indexed
	for ln in range(m):
		x, y, a, b = map(int, fin.readline().split())
		graph[(x, y)] = graph.get((x, y), []) + [(a, b)]

lit = {(1, 1)}
seen = set()
directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
q = deque([(1, 1)])

# Breadth First Search (BFS) lit rooms
while q:
	r, c = q.popleft()
	if (r, c) in seen:
		continue
	seen.add((r, c))

	# Turn on new lights of current room
	if (r, c) in graph:
		for sr, sc in graph[(r, c)]:
			if (sr, sc) in lit:
				continue
			lit.add((sr, sc))
			# Check if the newly lit room can be connected and hasn't been visited
			q += [(sr, sc) for dr, dc in directions if (sr + dr, sc + dc) in seen and (sr, sc) not in seen]

	# Explore valid adjacent rooms
	q += [(r + dr, c + dc) for dr, dc in directions if (r + dr, c + dc) in lit and (r + dr, c + dc) not in seen]

# Write number of lit rooms accessible from (1, 1) to output file
print(len(lit), file=open("lightson.out", "w"))

'''
# Graphs - Depth First Search (DFS) | Flood fill approach
# T & M: O(n^2 + m) where n is the number of rooms and m is the number of switches

with open("lightson.in", "r") as fin:
	# First line: n = # of rooms, m = # of switches
	n, m = map(int, fin.readline().split())
	# For the next m lines map x, y room to room a, b that can be lit
	graph = {}	# 1-indexed
	for ln in range(m):
		x, y, a, b = map(int, fin.readline().split())
		graph[(x, y)] = graph.get((x, y), []) + [(a, b)]

lit = {(1, 1)}
seen = set()
directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
stack = [(1, 1)]

# Iteratively Depth First Search (DFS) lit rooms
while stack:
	r, c = stack.pop()
	if (r, c) in seen:
		continue
	seen.add((r, c))

	# Turn on new lights of current room
	if (r, c) in graph:
		for sr, sc in graph[(r, c)]:
			lit.add((sr, sc))
			# Check if the newly lit room can be connected and hasn't been visited
			stack += [(sr, sc) for dr, dc in directions if (sr + dr, sc + dc) in seen and (sr, sc) not in seen]

	# Explore valid adjacent rooms
	stack += [(r + dr, c + dc) for dr, dc in directions if (r + dr, c + dc) in lit and (r + dr, c + dc) not in seen]

# Write number of lit rooms accessible from (1, 1) to output file
print(len(lit), file=open("lightson.out", "w"))
'''

File: Graphs/Silver/P1_2015-Switching_on_the_Lights/P1_2015-Switching_on_the_Lights.cpp
// Graphs - Breadth First Search (DFS) | Flood fill approach
// T & M: O(n^2 + m) 
// Where n is the number of rows and columns in the switches graph and m is the number of switches

#include <bits/stdc++.h>

using namespace std;

struct indices {const int row, col;};

int main() {
	freopen("lightson.in", "r", stdin);
	// First line: n = graph side length, m = # of switches
	int n, m;
	cin >> n >> m;
	// For the next m lines map x, y room to room a, b that can be lit
	vector<vector<vector<indices>>> switches(n, vector<vector<indices>>(n));
	for (int ln = 0; ln < m; ++ln) {
		int x, y, a, b;
		cin >> x >> y >> a >> b;
		switches[x - 1][y - 1].push_back({a - 1, b - 1});
	}

	vector<indices> directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
	vector<vector<bool>> seen(n, vector<bool>(n)), lit(n, vector<bool>(n));
	queue<indices> q;

	int litCount = 1;
	lit[0][0] = true;
	q.push({0, 0});
	while (!q.empty()) {
		indices qd = q.front();
		q.pop();
		if (seen[qd.row][qd.col]) 
			continue;
		seen[qd.row][qd.col] = true;

		// Turn on new lights of current room
		for (indices& sw : switches[qd.row][qd.col]) {
			if (lit[sw.row][sw.col])
				continue;
			lit[sw.row][sw.col] = true;
			++litCount;
			// Check if the newly lit room can be connected and hasn't been visited
			for (indices& d : directions) {
				const int aswr = sw.row + d.row, aswc = sw.col + d.col;
				if ((-1 < aswr && aswr < n) && 
				(-1 < aswc && aswc < n) && 
				seen[aswr][aswc] &&
				!seen[sw.row][sw.col])
					q.push(sw);
			}
		}

		// Explore valid adjacent rooms
		for (indices& d : directions) {
			const int ar = qd.row + d.row, ac = qd.col + d.col;
			if ((-1 < ar && ar < n) && 
			(-1 < ac && ac < n) &&
			lit[ar][ac] &&
			!seen[ar][ac])
				q.push({ar, ac});
		}
	}

	// Write number of lit rooms accessible from (1, 1) to output file
	freopen("lightson.out", "w", stdout);
	cout << litCount << endl;
	return 0;
}